CONTRIBUTING.md

```diff
@@ -10,55 +10,145 @@ For general contribution and community guidelines, please see the [community rep
 
 ## Development
 
+### Building and packaging
+
+To build versions for Linux, macOS and Windows:
+
+```
+./build
+```
+
+Binaries will be placed in `dist/`.
+
+### Running the project from source
+
 Run the project with:
 
 ```
 go run cmd/main.go
 ```
 
-### Testing
+### Unit Testing
 
 Tests are written using [GoConvey](http://goconvey.co/).
 Run tests with `go test -v ./...` or `./test` (for CI).
 
-### Building and packaging
+### Smoke Testing
 
-To build versions for Linux, macOS and Windows:
+Smoke testing is used to verify that the output generated by our `./build` script
+functions as expected.
+
+This is done through running the generated binaries against production or custom
+providers and verifying the output manually.  
+
+### Generating Mock Providers
+
+For testing purposes, it is often necessary to create a "mock" provider which can verify 
+input and output. This can be done by creating a script which can be run by Summon and
+placing it in the default path, `/usr/local/lib/summon/`
 
+For example, we can create a provider that simply returns the literal variable that is
+passed into it:
+
+1. Create a shell script in your Summon path called `mock`:
+```bash
+sudo <nano, vim, etc.> /usr/local/lib/summon/mock
 ```
-./build
+
+2. Add the following to your script:
+```bash
+#!/bin/bash -e
+echo "$@"
```

Comment:
You can do this in a single `echo` line or `cat`

Reply:
Sorry, I think I'm missing what you're suggesting. Can you give an example of what you mean?

Reply:
Good point! I'll simplify the commands into a single one.